pipeline/utils/github_utils.py

```diff
@@ -0,0 +1,165 @@
+    if attempt.ok:
+        return json.loads(attempt.content)
+    else:
+        raise HTTPError(url, attempt.status_code, attempt.content,
+                        attempt.headers, None)
+
+
+def _get_file_mode(el):
+    if os.path.isfile(el) and os.access(el, os.X_OK):
+        return _EXEC_MODE
+    return _FILE_MODE
+
+
+def _get_subtree_sha(orig_tree, path):
+    gen = (x for x in orig_tree['tree'] if (x['path'] == path and
+                                            x['mode'] == _TREE_MODE))
+    subtree_item = next(gen, None)
+    if subtree_item:
+        return subtree_item['sha']
+
+
+# TODO: Have this function also upload symlinks to the repository
+# according to GitHub protocol
+def _extend_git_tree(curr_dir, username, password, base_url,
+                     orig_tree_sha=None):
+    """Create a git tree from contents of local directory curr_dir.
+
+    If orig_tree_sha is provided, the new tree inherits all structure from
+    the remote tree with sha orig_tree_sha. The new tree will always include
+    files or paths that are present in the remote tree. However, if the
+    curr_dir and the remote tree  have conflicting contents for files with the
+    same path, the new tree will use the content of curr_dir.
+
+    Returns:
+        New tree object as a dictionary, if curr_dir had files
+        Otherwise, returns None
+
+    Raises:
+        HTTPError: An HTTP request used to construct the tree failed
+
+    """
+    if orig_tree_sha:
+        orig_tree = _exec_request(
+            requests.get, base_url + 'git/trees/{0}'.format(orig_tree_sha))
+    tree_els = []
+    for subdir in _get_immediate_subdirs(curr_dir):
+        subtree_sha = None
+        if orig_tree_sha:
+            subtree_sha = _get_subtree_sha(orig_tree, subdir)
+        subtree = _extend_git_tree(os.path.join(curr_dir, subdir), username,
+                                   password, base_url, subtree_sha)
+        # Add the subtree to tree_els iff subdir contains files
+        if subtree:
+            tree_els.append({'path': subdir, 'mode': _TREE_MODE,
+                             'type': 'tree', 'sha': subtree['sha']})
+    for exec_or_file in _get_immediate_files(curr_dir):
+        with open(os.path.join(curr_dir, exec_or_file), 'r') as f:
+            path_content = f.read()
+        tree_els.append({'path': exec_or_file,
+                         'mode': _get_file_mode(os.path.join(curr_dir,
+                                                             exec_or_file)),
+                         'type': 'blob', 'content': path_content})
+    # Git doesn't allow empty trees
+    if tree_els:
+        data_dict = {'tree': tree_els}
+        if orig_tree_sha:
+            data_dict['base_tree'] = orig_tree_sha
+        return _exec_request(requests.post, base_url + 'git/trees',
+                             json.dumps(data_dict), username, password)
+
+
+def push_dir_github(output_dir, username, password, owner, repo, branch,
```

Comment:
Since dir is reserved name in python, renamed output_dir to local_dir. Renamed push_dir_github to push_dir_to_github.